### PR TITLE
Top Spin Track paint

### PR DESCRIPTION
**Pull request by duncanspumpkin**
This implements all of the top spin paint setup functions. I've also had to make modifications to the track paint setup code so that it can use the new functions.

Can someone with the linux native build check this out to see if #2063 happens with this version.

### CODE REVIEW

File: src/ride/track_paint.h [hunk 1]
```diff
@@ -0,0 +1,10 @@
+#ifndef _TRACK_PAINT_H
+#define _TRACK_PAINT_H
+
+#include "../common.h"
+
+typedef void (TRACK_PAINT_FUNCTION)(uint8 rideIndex, uint8 trackSequence, uint8 direction, int height, rct_map_element* mapElement);
```
Comment (IntelOrca) [inline src/ride/track_paint.h]: We usually pre-typedef the pointer so we don't have to do in each usage.

File: src/interface/viewport.c [hunk 2]
```diff
@@ -1398,28 +1411,34 @@ void viewport_track_paint_setup(uint8 direction, int height, rct_map_element *ma
...
-			direction,
-			height,
-			(int)mapElement,
-			rideIndex * sizeof(rct_ride),
-			trackSequence
-		);
+		TRACK_PAINT_FUNCTION ***trackTypeList = (TRACK_PAINT_FUNCTION***)RideTypeTrackPaintFunctionsOld[ride->type];
+		if (trackTypeList == NULL) {
+			trackTypeList = (TRACK_PAINT_FUNCTION***)RideTypeTrackPaintFunctions[ride->type];
+			trackTypeList[trackType][direction](rideIndex, trackSequence, direction, height, mapElement);
```
Comment (IntelOrca) [inline src/interface/viewport.c]: If we fetch this and then check for NULL here, we can eliminate using the dummy function.

File: src/ride/track_paint.h (hunk 1 above)
Comment (duncanspumpkin) [inline src/ride/track_paint.h]: Not sure what you mean by pre-typedef.

**Comment by IntelOrca** [comment]
For top_spin_paint_setup_rot_0, top_spin_paint_setup_rot_1, top_spin_paint_setup_rot_2 and top_spin_paint_setup_rot_3... we could pre-define arrays of Z tile order.

File: src/ride/track_paint.h (hunk 1 above)
Comment (IntelOrca) [inline src/ride/track_paint.h]: typedef void (TRACK_PAINT_FUNCTION*)(uint8 rideIndex, uint8 trackSequence, uint8 direction, int height, rct_map_element* mapElement);

**Comment by HaasJona** [comment]
The top spin animation is still broken with latest develop on linux.

**Comment by janisozaur** [comment]
True, top spin does not work correctly yet, but now there are things you can see with all the [inlines disabled](https://github.com/janisozaur/OpenRCT2/tree/ppc). That's still progress, so good work!

**Pull request merged**